### ISSUE TITLE
Annotation classloading issue in custom JUnit extension when run in continuous testing

### ISSUE THREAD

**Issue by famod** [quarkusio/quarkus#24257]
### Describe the bug

A Junit extension, e.g. a `org.junit.jupiter.api.extension.BeforeEachCallback`, that is registered via `src/test/resources/META-INF/services/org.junit.jupiter.api.extension.Extension`, cannot find a custom annotation that is present on a test method.

`context.getRequiredTestMethod().getAnnotationsByType(SomeAnnotation.class)` (`context` being `org.junit.jupiter.api.extension.ExtensionContext`) returns an empty array.
OTOH, `context.getRequiredTestMethod().getAnnotations()` does contain the annotation. So it's there, but the type is loaded in two different classloaders:

Via surefire (where everything is ok):
```
::: CL1: jdk.internal.loader.ClassLoaders$AppClassLoader@277050dc
::: CL2: jdk.internal.loader.ClassLoaders$AppClassLoader@277050dc
```
(CL1 is `SomeAnnotation.class.getClassLoader()` and CL2 is `annotation.annotationType().getClassLoader()`)

But in CT:
```
::: CL1: QuarkusClassLoader:Augmentation Class Loader: TEST@7497f2dc
::: CL2: QuarkusClassLoader:Deployment Class Loader: TEST@61ad719b
```

ℹ️ There is no issue when the extension is registered via `@ExtendWith(...)`!

### Expected behavior

Annotation is found in CT just like when run via surefire.

### Actual behavior

Annotation is not found by type because two different classloaders are involved.

### How to Reproduce?

[q_ct-callback-ann_multi.zip](https://github.com/quarkusio/quarkus/files/8227048/q_ct-callback-ann_multi.zip)

- `mvn clean install` ✔️ 
- `mvn quarkus:test -f dist` ❌ 

Note that this reproducer needs two modules because of #24229 and in reality, `core` produces a `test-jar` but I wanted to take that out of the equation.

_Edit: Udpated reproducer is here: https://github.com/quarkusio/quarkus/issues/24257#issuecomment-1076922903_

### Output of `java -version`

17.0.2, vendor: BellSoft

### Quarkus version or git rev

2.7.3.Final and 2.7.4.Final

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.8.4

### Additional information

Reminds me a lot of #10623, but that was long before CT was invented.

**Comment by aloubyansky** [comment]
@famod this appears to be fixed in main (but fails in 2.7.5.Final). I suppose https://github.com/quarkusio/quarkus/pull/24440 fixed it.
I did see your comment https://github.com/quarkusio/quarkus/issues/24229#issuecomment-1073374087, are you sure it's still an issue?

**Comment by famod** [comment]
@aloubyansky well, I checked it with my real project while trying out #24440 and it wasn't fixed.

I'll try again with current main. Maybe it's now fixed for the reproducer but not my real project (I hope not).

**Comment by aloubyansky** [comment]
Please do and see if you can adjust the reproducer for main. Thanks!

**Comment by famod** [comment]
I can confirm that it's fixed for the reproducer, but not for my real project. I'm somehow suspecting `test-jar`, will try to reproduce.

**Comment by famod** [comment]
It's not `test-jar` (at least it works in the small reproducer). Great, once again I have to find the needle in the haystack...

**Comment by famod** [comment]
Interesting: In my project where it fails (it doesn't fail everywhere!) there *two different* deployment classloaders involved!
In cases where it's fine it's the same deployment CL, just like in the simple reproducer.

Will try to dig deeper tomorrow, but in cas eyou have an idea @aloubyansky please let me know! 🙂

**Comment by famod** [comment]
One more detail: Right before the test there is a `@QuarkusTest` (with a `@QuarkusTestResource`, hence triggering a Quarkus restart) and the test that's failing ist a *non-Quarkus* test!

**Comment by famod** [comment]
Managed to reproduce it, it was just a matter of adding a non-Quarkus test: [q_ct-callback-ann_multi-2.zip](https://github.com/quarkusio/quarkus/files/8337304/q_ct-callback-ann_multi-2.zip)

**Comment by famod** [comment]
I suppose the deployment classloaders mismatch is caused by the creation of a new deployment classloader during discovery of unit test classes:
https://github.com/quarkusio/quarkus/blob/619288dfa9b11ee6591b8beace4810abef4ac468/core/deployment/src/main/java/io/quarkus/deployment/dev/testing/JunitTestRunner.java#L667-L671

/cc @stuartwdouglas @aloubyansky

**Comment by famod** [comment]
I think I have a fix for this, stay tuned!

**Comment by famod** [comment]
@stuartwdouglas @aloubyansky
I'm kinda stuck. This is what I came up with: https://github.com/quarkusio/quarkus/compare/main...famod:ct-unit-transform-cl
Which does resolves this issue, but it somehow also partially breaks change detection for QuarkusTest, meaning a QuarkusTest will no re-run if a class it depends on changes in an upstream module.

I'm under the impression that unit tests must be run separately, with a separate `TestsPlan`, with that special deployment classloader as TCCL...?

**Comment by famod** [comment]
PS: I noticed that reloading of unit tests is partially broken the same way, even in 2.7.5.Final.

**Comment by aloubyansky** [comment]
@famod just tried your last reproducer from https://github.com/quarkusio/quarkus/issues/24257#issuecomment-1076922903. It fails with 2.7.5.Final but passes with 2.8.0.Final and the main branch.

**Comment by famod** [comment]
@aloubyansky hum, still fails for me with 2.8.0.Final and main (on two different machines):
```
mvn clean install && mvn quarkus:test -f dist
```

**Comment by aloubyansky** [comment]
Right. Not sure what happened. Just tried on another laptop and it fails. Sorry.

**Comment by famod** [comment]
No worries, thanks for checking!

**Comment by holly-cummins** [comment]
I wonder if this is also related to https://github.com/quarkusio/quarkus/issues/27821?

**Comment by famod** [comment]
@holly-cummins thanks for the pointer! At first glance #27821 seems to be about testing in general, not just continuous testing, right?
Thing is that this issue here only manifests in CT.

**Comment by holly-cummins** [comment]
Oh, good point. *Most* of my pain is related to continuous testing, but so I assumed https://github.com/quarkusio/quarkus/issues/27821 also was, but the detailed description says it affects both modes. So it will be distantly related, if it's related.

**Comment by holly-cummins** [comment]
@famod, is this still an issue on the latest 3.31 Quarkus?

**Comment by geoand** [comment]
Closing for lack of feedback
